packages/taco/src/conditions/base/shared.ts

```diff
@@ -1,11 +1,44 @@
 import { z } from 'zod';
 
-import { ETH_ADDRESS_REGEXP, USER_ADDRESS_PARAM } from '../const';
+import {
+  CONTEXT_PARAM_REGEXP,
+  ETH_ADDRESS_REGEXP,
+  USER_ADDRESS_PARAM,
+} from '../const';
+import { CONTEXT_PARAM_PREFIX } from '../context/context';
+
+export const contextParamSchema = z.string().regex(CONTEXT_PARAM_REGEXP);
+// We want to discriminate between ContextParams and plain strings
+// If a string starts with `:`, it's a ContextParam
+export const plainStringSchema = z.string().refine(
+  (str) => {
+    return !str.startsWith(CONTEXT_PARAM_PREFIX);
+  },
+  {
+    message: 'Context parameters must start with ":"',
+  },
+);
+
+export const paramPrimitiveTypeSchema = z.union([
+  plainStringSchema,
+  z.boolean(),
+  z.number(),
+]);
+export const paramSchema = z.union([
+  paramPrimitiveTypeSchema,
+  z.array(paramPrimitiveTypeSchema),
+]);
+
+export const paramOrContextParamSchema: z.ZodSchema = z.union([
+  paramSchema,
+  contextParamSchema,
+  z.lazy(() => z.array(paramOrContextParamSchema)),
+]);
```

Comment:
Are we being redundant here, with `paramPrimitivetypesSchema`, `paramSchema` and `paramOrContextParamSchema`?

It's just unclear to me that we need to separate paramSchema from paramOrContextParamSchema - they feel like one in the same, and won't be used separately. Can we achieve the same thing that boils down to the following (pseudocode)? 

```
strings_that_dont_start_with_colon = z.string().refine(
(str) => {
    return !str.startsWith(CONTEXT_PARAM_PREFIX);
  },
  {
    message: "String must not be a context parameter i.e. not start with ':'",
   })

paramPrimitiveTypesSchema = union[
     strings_that_dont_start_with_colon,
     strings_that_are_context_params,
     boolean,
     number
]

paramSchema = union[
     paramPrimitiveTypesSchema
     lazy -> array(paramPrimitiveTypesSchema)
]
```

NOTE: The names used above (`strings_that_..`) is not a recommendation for name change 😅 , just simplifying the picture for myself and potentially others. The original names are fine.

Reply:
Ah, yes, that's one-too-many schemas